scene/animation/animation_tree.h

```diff
@@ -54,6 +54,12 @@ class AnimationNode : public Resource {
 		FILTER_BLEND
 	};
 
+	enum AnimationNodeNotification {
+		ANIMATION_NODE_NOTIFICATION_STARTED,
+		ANIMATION_NODE_NOTIFICATION_FINISHED,
+		ANIMATION_NODE_NOTIFICATION_FADEOUT_STARTED
+	};
+
```

Comment:
I think the overall implementation is headed in the right direction.

My concern is where to define the enum for signal, maybe it should be defined separately for each AnimationNode and returned as an int in the AnimationTree like:

animation_blend_tree.h
```
	enum AnimationNodeNotificationOneShot {
		ANIMATION_NODE_NOTIFICATION_ONESHOT_STARTED,
		ANIMATION_NODE_NOTIFICATION_ONESHOT_FINISHED,
		ANIMATION_NODE_NOTIFICATION_FADEOUT_FINISHED_INTERNAL, // Fadeout started.
	};
~~~~~~
	enum AnimationNodeTransitionNotification {
		ANIMATION_NODE_NOTIFICATION_TRANSITION_STARTED,
		ANIMATION_NODE_NOTIFICATION_TRANSITION_FINISHED,
	};
```

Reply:
Yeah I was thinking about this. The meaning of some of the notifications are frustratingly similar but I agree it's probably best to separate them for clarity and ease of documentation.

Reply:
Revisiting this, I think actually each of the oneshot, state machine, and transition nodes could use `STARTED`, `FINISHED`, and `FINISHED_INTERNAL` with near identical semantics. At least for next version of this pr I'm planning on having those three constants defined on AnimationNode but use int for the signal/helper method so that AnimationNode subclasses can still define their own notification values and document them individually.

I do have a question as this version of the pr has `FADEOUT_STARTED` as opposed to `FINISHED_INTERNAL` and I've kept the code to only emit that value when there is nonzero crossfade; would you want `FINISHED_INTERNAL` to fire in cases where there is no fadeout?

Reply:
I am beginning to think that FINISHED_INTERNAL is a little too difficult for users who do not touch the core, so FADEOUT_STARTED might be fine.

> I do have a question as this version of the pr has FADEOUT_STARTED as opposed to FINISHED_INTERNAL and I've kept the code to only emit that value when there is nonzero crossfade; would you want FINISHED_INTERNAL to fire in cases where there is no fadeout?

If there is no crossfade, it is better not to fire. However, if the implementation is too difficult, it is fine to fire it, but perhaps you should note that it fires even if there is no fade-out in the documentation.

Reply:
Ok, I had to have this on the backburner for a bit but I've updated the pr with per-class enums. I ended up using unnamed enums with manually-made-unique values to be more similar to notification but that can easily be changed to whatever.

I think my only issue with the current implementation is how verbose qualifying the enum is but maybe either the name could be changed or it could be bound as a new globalscope constant. I also noticed a while ago that a state machine with no valid path to the end node never actually finishes (existing behavior; a one shot triggering such a sm will not finish) and included that in the documentation although I'm not sure if it's really the desired behavior?